modules/azure/compute.go

```diff
@@ -8,97 +8,444 @@ import (
+	// Get VM Object
+	vm, err := GetVirtualMachineE(vmName, resGroupName, subscriptionID)
 	if err != nil {
 		return "", err
 	}
 
-	// Create a VM client
-	vmClient, err := GetVirtualMachineClientE(subscriptionID)
+	return *vm.StorageProfile.OsDisk.Name, nil
+}
+
+// GetVirtualMachineState gets the State of the specified Azure Virtual Machine.
+// This function would fail the test if there is an error.
+func GetVirtualMachineState(t testing.TestingT, vmName string, resGroupName string, subscriptionID string) string {
+	vmState, err := GetVirtualMachineStateE(vmName, resGroupName, subscriptionID)
+	require.NoError(t, err)
+	return vmState
+}
+
+// GetVirtualMachineStateE gets the State of the specified Azure Virtual Machine.
+func GetVirtualMachineStateE(vmName string, resGroupName string, subscriptionID string) (string, error) {
+	// Get VM Object
+	vm, err := GetVirtualMachineE(vmName, resGroupName, subscriptionID)
+	if err != nil {
+		return "", err
+	}
+
+	return vm.Status, nil
+}
+
+// GetVirtualMachineAvailabilitySetID gets the Availability Set ID of the specified Azure Virtual Machine.
+// This function would fail the test if there is an error.
+func GetVirtualMachineAvailabilitySetID(t testing.TestingT, vmName string, resGroupName string, subscriptionID string) string {
+	avsID, err := GetVirtualMachineAvailabilitySetIDE(vmName, resGroupName, subscriptionID)
+	require.NoError(t, err)
+
+	return avsID
+}
+
+// GetVirtualMachineAvailabilitySetIDE gets the Availability Set ID of the specified Azure Virtual Machine.
+func GetVirtualMachineAvailabilitySetIDE(vmName string, resGroupName string, subscriptionID string) (string, error) {
+	// Get VM Object
+	vm, err := GetVirtualMachineE(vmName, resGroupName, subscriptionID)
+	if err != nil {
+		return "", err
+	}
+
+	if vm.AvailabilitySet == nil {
+		return "", NewNotFoundError("Availability Set", "", vmName)
+	}
+
+	return GetNameFromResourceID(*vm.AvailabilitySet.ID), nil
+}
+
+// VMImage represents the storage image for the specified Azure Virtual Machine.
+type VMImage struct {
+	Publisher string
+	Offer     string
+	SKU       string
+	Version   string
+}
+
+// GetVirtualMachineImage gets the Image of the specified Azure Virtual Machine.
+// This function would fail the test if there is an error.
+func GetVirtualMachineImage(t testing.TestingT, vmName string, resGroupName string, subscriptionID string) VMImage {
+	adminUser, err := GetVirtualMachineImageE(vmName, resGroupName, subscriptionID)
+	require.NoError(t, err)
+
+	return adminUser
+}
+
+// GetVirtualMachineImageE gets the Image  of the specified Azure Virtual Machine.
+func GetVirtualMachineImageE(vmName string, resGroupName string, subscriptionID string) (VMImage, error) {
+	var vmImage VMImage
+
+	// Get VM Object
+	vm, err := GetVirtualMachineE(vmName, resGroupName, subscriptionID)
+	if err != nil {
+		return vmImage, err
+	}
+
+	if vm.StorageProfile == nil {
+		return vmImage, NewNotFoundError("Image Reference", "Any", vmName)
```

Comment:
I think the string in this function in errors. go is wrong, your third param is not a region as it should not be.  Should we refactor in this PR or seperate?

func (err NotFoundError) Error() string {
	return fmt.Sprintf("Object of type %s with id %s not found in region %s", err.objectType, err.objectID, err.region)
}

Reply:
There is a NewNotFoundError now also in modules/azure/errors.go, this one is more generic.